pkg/gadgets/internal/socketenricher/bpf/sockets-map.bpf.c

```diff
@@ -0,0 +1,231 @@
+// SPDX-License-Identifier: (GPL-2.0 WITH Linux-syscall-note) OR Apache-2.0
+/* Copyright (c) 2021 The Inspektor Gadget authors */
+
+#include <vmlinux/vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_endian.h>
+
+#include <bpf/bpf_helpers.h>
+
+#define SOCKETS_MAP_IMPLEMENTATION
+#include "sockets-map.h"
+
+#define MAX_ENTRIES	10240
+
+// The map 'start' keeps context between a kprobe and a kretprobe
+// Keys: pid_tgid
+// Values: the argument of the kprobe function:
+// - When used in bind: struct socket *
+// - When used in tcp_connect: struct sock *
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, __u64);
+	__type(value, void *);
+} start SEC(".maps");
+
+static __always_inline void
+update_sockets_map(struct sock *sock)
+{
+	struct sockets_key socket_key = {0,};
+	struct inet_sock *inet_sock = (struct inet_sock *)sock;
+	BPF_CORE_READ_INTO(&socket_key.netns, sock, __sk_common.skc_net.net, ns.inum);
+	socket_key.proto = BPF_CORE_READ(sock, sk_protocol);
+	socket_key.port = bpf_ntohs(BPF_CORE_READ(inet_sock, inet_sport));
+
+	struct sockets_value socket_value = {0,};
+	struct task_struct *task = (struct task_struct*) bpf_get_current_task();
+	socket_value.mntns = (u64) BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
+	socket_value.pid_tgid = bpf_get_current_pid_tgid();
+	bpf_get_current_comm(&socket_value.task, sizeof(socket_value.task));
+
+	bpf_map_update_elem(&sockets, &socket_key, &socket_value, BPF_ANY);
```

Comment:
Do you mean you prefer to give no information than potentially wrong information if the key is not unique?

Reply:
Exactly.

Reply:
I am not sure which behaviour is more helpful. Do you know of applications using SO_REUSEPORT on Kubernetes? Since the netns is in the key, at least the 2 processes reusing the port would be in the same netns. But it could be different pods both using host networking. I'd like to have an example of such case to see if it's more practical for the user to have an idea of the application, or if it's better to not give any indications.

Reply:
What about checking if `SO_REUSEPORT` is enabled?, if it's then we can replace the element, otherwise we shouldn't replace the element because for some reason there was a clash. 
I also think this is not that important if we include family and source address in the key since I don't see other possibilities for collisions to happen other than `SO_REUSEPORT`.